### PR TITLE
[Feature] Support disk disable/decommission (part1)

### PR DESCRIPTION
**Pull request by gengjun-git**
Why I'm doing:
1. In the case of disk damage, currently the tablet can only be repaired by modifying the disk configuration and restarting BE. However, for some customers, restarting BE is unacceptable.
2. For some reasons, such as disk performance, customers need to replace a disk. Currently, this is done by changing the configuration and restarting BE. Moreover, disk migration cannot yet be performed for single-replica tables.

What I'm doing:
1. Support disable disk: users can send a disable disk command: tablets on that disk will be removed, and a new replica of that tablet will be copied to health disks.
2. Support decommission disk: users can send a decommission disk command: tablets on that disk will be migrated to other disks on the same BE.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -964,6 +968,22 @@ cleanTabletSchedQClause
...
+decommissionDiskClause
+    : DECOMMISSION DISK string (',' string)*
+    ;
+
+cancelDecommissionDiskClause
+    : CANCEL DECOMMISSION DISK string (',' string)*
+    ;
+
+disableDiskClause
+    : DISABLE DISK string (',' string)*
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:980]: Fot not implemented sql interface, we should return an unsupported error message when user executes that sql.(In EE version, we don't.)

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 2]
```diff
@@ -964,6 +968,22 @@ cleanTabletSchedQClause
     : CLEAN TABLET SCHEDULER QUEUE
     ;
 
+decommissionDiskClause
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:971]: Should make the disk decommission and disk disable cannot run simutaneourly.

**Review by nshangyiming — COMMENTED** [review]
TBD

File: fe/fe-core/src/main/java/com/starrocks/catalog/LocalTablet.java [hunk 3]
```diff
@@ -696,12 +704,17 @@ private Pair<TabletStatus, TabletSchedCtx.Priority> getHealthStatusWithPriorityU
...
         if (replicas.size() > replicationNum) {
             // we set REDUNDANT as VERY_HIGH, because delete redundant replicas can free the space quickly.
-            return createRedundantSchedCtx(TabletStatus.REDUNDANT, TabletSchedCtx.Priority.VERY_HIGH,
+            return createRedundantSchedCtx(TabletStatus.REDUNDANT, Priority.VERY_HIGH,
                     needFurtherRepairReplica);
         }
 
-        // 5. healthy
-        return Pair.create(TabletStatus.HEALTHY, TabletSchedCtx.Priority.NORMAL);
+        // 5. disk decommission
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/catalog/LocalTablet.java:711]: why don't put disk decommission before `4. replica redundant`?

File: be/src/storage/data_dir.cpp [hunk 4]
```diff
@@ -151,14 +150,24 @@ Status DataDir::set_cluster_id(int32_t cluster_id) {
 
 void DataDir::health_check() {
     // check disk
-    if (_is_used) {
-        Status res = _read_and_write_test_file();
-        if (!res.ok()) {
-            LOG(WARNING) << "store read/write test file occur IO Error. path=" << _path << ", res=" << res.to_string();
-            if (is_io_error(res)) {
-                _is_used = false;
+    if (_state != DiskState::DECOMMISSIONED && _state != DiskState::DISABLED) {
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/data_dir.cpp:153]: There is another state: OFFLINE

File: be/src/storage/data_dir.h [hunk 5]
```diff
@@ -56,6 +56,13 @@ class Tablet;
 class TabletManager;
 class TxnManager;
 
+enum DiskState {
+    ONLINE,
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/data_dir.h:60]: This is disk state, not tablet state

File: be/src/storage/storage_engine.cpp [hunk 6]
```diff
@@ -1541,4 +1541,33 @@ void StorageEngine::clear_rowset_delta_column_group_cache(const Rowset& rowset)
     }());
 }
 
+void StorageEngine::disable_disks(const std::vector<string>& disabled_disks) {
+    std::lock_guard<std::mutex> l(_store_lock);
+
+    for (auto& it : _store_map) {
+        if (std::find(disabled_disks.begin(), disabled_disks.end(), it.second->path()) != disabled_disks.end()) {
+            it.second->set_state(DiskState::DISABLED);
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/storage_engine.cpp:1549]: done

File: be/src/storage/storage_engine.cpp [hunk 7]
```diff
@@ -1541,4 +1541,33 @@ void StorageEngine::clear_rowset_delta_column_group_cache(const Rowset& rowset)
...
+        } else {
+            if (it.second->get_state() == DiskState::DISABLED) {
+                it.second->set_state(DiskState::ONLINE);
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/storage_engine.cpp:1553]: done

File: be/src/storage/storage_engine.cpp [hunk 8]
```diff
@@ -1541,4 +1541,33 @@ void StorageEngine::clear_rowset_delta_column_group_cache(const Rowset& rowset)
...
+    }
+}
+
+void StorageEngine::decommission_disks(const std::vector<string>& decommission_disks) {
+    std::lock_guard<std::mutex> l(_store_lock);
+
+    for (auto& it : _store_map) {
+        if (std::find(decommission_disks.begin(), decommission_disks.end(), it.second->path()) !=
+            decommission_disks.end()) {
+            it.second->set_state(DiskState::DECOMMISSIONED);
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/storage_engine.cpp:1566]: done

File: fe/fe-core/src/main/java/com/starrocks/catalog/DiskInfo.java [hunk 9]
```diff
@@ -49,7 +49,13 @@
 public class DiskInfo implements Writable {
     public enum DiskState {
         ONLINE,
-        OFFLINE
+        // Reported from BE, if disk is detected as unavailable, state will be OFFLINE.
+        // Tablets on OFFLINE disk will be dropped.
+        OFFLINE,
+        // Set by user, tablets on DISABLED disk will be dropped.
+        DISABLED,
+        // Set by user, tablets on DECOMMISSIONED disk will be cloned to other backends.
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/catalog/DiskInfo.java:57]: done

File: fe/fe-core/src/main/java/com/starrocks/system/Backend.java [hunk 10]
```diff
@@ -276,6 +295,34 @@ public void updateDisks(Map<String, TDisk> backendDisks) {
         }
     }
 
+    public void decommissionDisk(String rootPath) {
+        DiskInfo diskInfo = disksRef.get(rootPath);
+        if (diskInfo != null) {
+            diskInfo.setState(DiskState.DECOMMISSIONED);
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/system/Backend.java:305]: done

File: be/src/storage/data_dir.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (nshangyiming) [reply be/src/storage/data_dir.h:60]: be_tablets will show data dir for every tablet, we can add this col later if needed i think

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 11]
```diff
@@ -964,6 +968,22 @@ cleanTabletSchedQClause
...
+
+cancelDecommissionDiskClause
+    : CANCEL DECOMMISSION DISK string (',' string)*
+    ;
+
+disableDiskClause
+    : DISABLE DISK string (',' string)*
+    ;
+
+cancelDisableDiskClause
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:983]: we should also have a syntax sugar like `show proc "/backends/*";`, otherwist it's quite cubersome to should the disk state of all backends.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 12]
```diff
@@ -964,6 +968,22 @@ cleanTabletSchedQClause
     : CLEAN TABLET SCHEDULER QUEUE
     ;
 
+decommissionDiskClause
+    : DECOMMISSION DISK string (',' string)*
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:972]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 1 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:980]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 2 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:971]: done

File: be/src/storage/data_dir.cpp [hunk 13]
```diff
@@ -151,14 +150,24 @@ Status DataDir::set_cluster_id(int32_t cluster_id) {
...
+        bool all_failed = true;
+        for (int i = 0; i < 10; i++) {
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/data_dir.cpp]: done

File: be/src/storage/data_dir.h [hunk 14]
```diff
@@ -56,6 +56,13 @@ class Tablet;
...
+    OFFLINE,       // detected by health_check, tablets on OFFLINE disk will be dropped.
+    DISABLED,      // set by user, tablets on DISABLED disk will be dropped.
+    DECOMMISSIONED // set by user, tablets on DECOMMISSIONED disk will be cloned to other backends.
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply be/src/storage/data_dir.h]: done

File: fe/fe-core/src/main/java/com/starrocks/catalog/LocalTablet.java (hunk 3 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/catalog/LocalTablet.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 15]
```diff
@@ -964,6 +968,22 @@ cleanTabletSchedQClause
...
+    ;
+
+cancelDecommissionDiskClause
+    : CANCEL DECOMMISSION DISK string (',' string)*
+    ;
+
+disableDiskClause
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:979]: The single replica will not lose.

File: fe/fe-core/src/main/java/com/starrocks/leader/ReportHandler.java [hunk 16]
```diff
@@ -890,6 +888,11 @@ private static void deleteFromMeta(ListMultimap<Long, Long> tabletDeleteFromMeta
             }
         } // end for dbs
 
+        if (deleteTablets.size() > 0) {
+            GlobalStateMgr.getCurrentState().getEditLog()
+                    .logBatchDeleteReplica(new BatchDeleteReplicaInfo(backendId, deleteTablets));
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/leader/ReportHandler.java:893]: need to put it in lock?

Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/leader/ReportHandler.java:893]: There is no need to put it in db lock. If the related meta is dropped, the replay code can cover this case.

**Review by nshangyiming — DISMISSED** [review]

**Comment by gengjun-git** [comment]
@mergify backport branch-3.1 branch-3.2

**Pull request merged**